<!DOCTYPE html><html><head><title>R: Annotated, Summary Statistics with chromosomal index RsIDs</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.css">
<script type="text/javascript">
const macros = { "\\R": "\\textsf{R}", "\\code": "\\texttt"};
function processMathHTML() {
    var l = document.getElementsByClassName('reqn');
    for (let e of l) { katex.render(e.textContent, e, { throwOnError: false, macros }); }
    return;
}</script>
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.js"
    onload="processMathHTML();"></script>
<link rel="stylesheet" type="text/css" href="style.css" />
</head><body><div class="container">
<div class="topbar"><a href="index.html">MiamiR Help</a><span class="crumb">/ Annotate_Data</span></div>
<div class="container"><div class="card">

<table style="width: 100%;"><tr><td>Annotate_Data</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2 id='Annotate_Data'>Annotated, Summary Statistics with chromosomal index RsIDs</h2>

<h3>Description</h3>

<p>Annotated, Summary Statistics with chromosomal index RsIDs
</p>


<h3>Usage</h3>

<pre><code class='language-R'>Annotate_Data(..., session = NULL)
</code></pre>


<h3>Arguments</h3>

<table>
<tr><td><code id="Annotate_Data_:_...">...</code></td>
<td>
<p>Shadow argument to detect inherited arguments passed and modified from the Single_Plot() function; defaults to NULL</p>
</td></tr>
<tr><td><code id="Annotate_Data_:_Data">Data</code></td>
<td>
<p>These are the summary statistics (either file path(s),environmental object(s) or a variable(s) containing such contents) to be plotted; defaults to NULL (can be a list for multiple runs)</p>
</td></tr>
<tr><td><code id="Annotate_Data_:_Chromosome_Column">Chromosome_Column</code></td>
<td>
<p>Manually specify chromosome column; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Annotate_Data_:_Position_Column">Position_Column</code></td>
<td>
<p>Manually specify chromosome genomic position column; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Annotate_Data_:_SNP_ID_Column">SNP_ID_Column</code></td>
<td>
<p>Manually specify SNP ID column; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Annotate_Data_:_PValue_Column">PValue_Column</code></td>
<td>
<p>Manually specify P Value column; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Annotate_Data_:_Reference_Allele_Column">Reference_Allele_Column</code></td>
<td>
<p>Manually specify reference allele column; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Annotate_Data_:_Effect_Allele_Column">Effect_Allele_Column</code></td>
<td>
<p>Manually specify effect allele column; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Annotate_Data_:_Verbose">Verbose</code></td>
<td>
<p>Prevent display of progress bar as function is running and instead show key milestone outputs/messages (mainly for debugging purposes); defaults to FALSE.</p>
</td></tr>
<tr><td><code id="Annotate_Data_:_Genome_Build">Genome_Build</code></td>
<td>
<p>Reference genome to base required RSID annotations around; defaults to grch38 (grch37 also allowed)</p>
</td></tr>
</table>


<h3>Value</h3>

<p>Annotated data frame with Lab column containing RsIDs for chromosome index SNPs is allocated to specified object and the resulting summary statistics can then be passed to other MiamiR functions
</p>


<h3>Examples</h3>

<pre><code class='language-R'> Annotated_Data &lt;- Annotate_Data(Data = Intelligence_Sum_Stats,  Genome_Build = "grch37")
</code></pre>


</div>
</div></div>
</body></html>
