<!DOCTYPE html><html><head><title>R: Annotated, Customised Forest Plots</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.css">
<script type="text/javascript">
const macros = { "\\R": "\\textsf{R}", "\\code": "\\texttt"};
function processMathHTML() {
    var l = document.getElementsByClassName('reqn');
    for (let e of l) { katex.render(e.textContent, e, { throwOnError: false, macros }); }
    return;
}</script>
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.js"
    onload="processMathHTML();"></script>
<link rel="stylesheet" type="text/css" href="style.css" />
<body>
<div class="topbar"><a href="index.html">MiamiR Help</a><span class="crumb">/ Forest_Plot</span></div>
<div class="container"><div class="card">

<table style="width: 100%;"><tr><td>Forest_Plot</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2 id='Forest_Plot'>Annotated, Customised Forest Plots</h2>

<h3>Description</h3>

<p>Annotated, Customised Forest Plots
</p>


<h3>Usage</h3>

<pre><code class='language-R'>Forest_Plot(Data, ..., session = NULL)
</code></pre>


<h3>Arguments</h3>

<table>
<tr><td><code id="Forest_Plot_:_Data">Data</code></td>
<td>
<p>These are the summary statistics (either file path(s), environmental object(s) or a variable(s) containing such contents) to be plotted; defaults to NULL</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Names">Names</code></td>
<td>
<p>These are a list of manual data set names to be applied to the legend text/key on the main plot - this also specifies the order down the page; defaults to NULL</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Data_Set_Colours">Data_Set_Colours</code></td>
<td>
<p>These are a list of manual data set colours allocated within the main plot - this also specifies the order down the page; defaults to NULL</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Model_Reference">Model_Reference</code></td>
<td>
<p>Manually specify where the statistics to be used derive from (either from GWAS or Models); defaults to FALSE - not a model</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Shapes">Shapes</code></td>
<td>
<p>These are a list of manual data set shapes allocated within the main plot - this also specifies the order down the page; defaults to NULL</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Null_Line_Colour">Null_Line_Colour</code></td>
<td>
<p>Colour of null line; defaults to &quot;red&quot;</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Null_Line_Type">Null_Line_Type</code></td>
<td>
<p>Type of null line; defaults to &quot;dashed&quot;</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_X_Axis_Title">X_Axis_Title</code></td>
<td>
<p>X axis title; defaults to NULL</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_X_Axis_Title_Size">X_Axis_Title_Size</code></td>
<td>
<p>X axis title size; defaults to 13</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_X_Axis_Label">X_Axis_Label</code></td>
<td>
<p>Display X axis title; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_X_Axis_Separation">X_Axis_Separation</code></td>
<td>
<p>Continuous value of the specific break separation to show on the X axis; defaults to NULL - leading to auto-calculation</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Strip_Colour">Strip_Colour</code></td>
<td>
<p>Colour of the strips highlighting the groups of covariates or SNPs under a certain category; defaults to NULL</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Strips">Strips</code></td>
<td>
<p>Display strips highlighting groups of interest; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Pre_Calculated_CIs">Pre_Calculated_CIs</code></td>
<td>
<p>Indicate that custom CIs are pre-calculated and available in datasets; defaults to FALSE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Legend_Title_Size">Legend_Title_Size</code></td>
<td>
<p>Legend title size; defaults to 13</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Legend_Text_Size">Legend_Text_Size</code></td>
<td>
<p>Legend text size; defaults to 11</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Legend_Title">Legend_Title</code></td>
<td>
<p>Title to display next to the legend; defaults to &quot;Data&quot;</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Left_Title">Left_Title</code></td>
<td>
<p>Title to display for the information in the grid text module on the left hand side of the plot; defaults to NULL</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Match_Allele_Direction">Match_Allele_Direction</code></td>
<td>
<p>Match the allele directions to avoid effect direction flips; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Match_Allele_Study">Match_Allele_Study</code></td>
<td>
<p>Data set used as reference point for effect allele direction; defaults to NULL - automatically assigned as first set</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Selected_SNPs">Selected_SNPs</code></td>
<td>
<p>A list of SNPs to be shown in the forest plot; defaults to NULL - automatically selected for from the summary statistics data set</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Selected_Covariates">Selected_Covariates</code></td>
<td>
<p>A list of covariates to be shown in the forest plot; defaults to NULL - automatically selected for from the munged model data set</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_X_Axis_Text_Resolution">X_Axis_Text_Resolution</code></td>
<td>
<p>Number of decimal places to display on X axis text; defaults to NULL</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Legend_On">Legend_On</code></td>
<td>
<p>Display the main plot legend; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_X_Axis_Text_Size">X_Axis_Text_Size</code></td>
<td>
<p>Size of the X axis text labels; defaults to 11</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Null_Buffer">Null_Buffer</code></td>
<td>
<p>Units of space to avoid for X axis text labels around the NULL point; defaults to 0.00</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Styles">Styles</code></td>
<td>
<p>These are a list of styles to be applied to the grid text on the right hand side of the main plot - this also specifies the order down the page; defaults to NULL and accepts &quot;underline&quot; or &quot;bold&quot; as deviations</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Missings">Missings</code></td>
<td>
<p>Simulate shape positions for data points which may be missing from specific data sets plotted; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Null_Line_Thickness">Null_Line_Thickness</code></td>
<td>
<p>Manually specify the thickness of the corresponding Null Line; defaults to 0.5</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Test_Statistic_Choice">Test_Statistic_Choice</code></td>
<td>
<p>Manually specify which type of test statistic output is desired, OR or BETA; defaults to BETA - regardless of input type either output may be specified</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Lead_Studies">Lead_Studies</code></td>
<td>
<p>Manually specify which study's lead SNPs should be plotted when Selected_SNPs is NULL; defaults to NULL - leads to index SNPs from any set to be plotted</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Separation_Distance">Separation_Distance</code></td>
<td>
<p>The minimum genomic distance between significant regions of interest/lead SNPs to be defined as distinct peaks, defined by P_threshold value when Selected_SNPs is NULL; defaults to 1e6</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_P_threshold">P_threshold</code></td>
<td>
<p>Maximum value of P (minimum LOG10P) allowed for a SNP to be classified as initiating detection as the lead SNP of a peak in the function, when Selected_SNPs is NULL; defaults to 5e-8</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Chromosomes">Chromosomes</code></td>
<td>
<p>Manually specify which chromosomes from the summary statistics to run the function on, when Selected_SNPs is NULL; defaults to NULL - leads to all chromosomes being processed</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Double_Label">Double_Label</code></td>
<td>
<p>Display a second level of information under individual grid text layers on the left hand side of the main plot; defaults to FALSE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Shape_Size">Shape_Size</code></td>
<td>
<p>Manually specify the size of the Shapes added on main plot; defaults to 4</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_CI_Line_Height">CI_Line_Height</code></td>
<td>
<p>Manually specify the vertical length of the CI lines added on main plot; defaults to 0.35</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_CI_Line_Size">CI_Line_Size</code></td>
<td>
<p>Manually specify the thickness of the CI lines added on main plot; defaults to 0.4</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Shape_Thickness">Shape_Thickness</code></td>
<td>
<p>Manually specify the thickness of the Shapes added on main plot; defaults to 1</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Grid_Thickness">Grid_Thickness</code></td>
<td>
<p>Manually specify the thickness of the grid lines, when Grid_Lines_On is TRUE; defaults to 0.25</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Grid_Lines_On">Grid_Lines_On</code></td>
<td>
<p>Show grid lines with specified settings around left and right hand side annotations; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_SE_Column">SE_Column</code></td>
<td>
<p>Manually specify SE of Test Statistic column; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_OR_Column">OR_Column</code></td>
<td>
<p>Manually specify OR Test Statistic column; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Chromosome_Column">Chromosome_Column</code></td>
<td>
<p>Manually specify chromosome column; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_BETA_Column">BETA_Column</code></td>
<td>
<p>Manually specify BETA Test Statistic column; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Position_Column">Position_Column</code></td>
<td>
<p>Manually specify chromosome genomic position column; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_SNP_ID_Column">SNP_ID_Column</code></td>
<td>
<p>Manually specify SNP ID column; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_PValue_Column">PValue_Column</code></td>
<td>
<p>Manually specify P Value column; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Reference_Allele_Column">Reference_Allele_Column</code></td>
<td>
<p>Manually specify reference allele column; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Effect_Allele_Column">Effect_Allele_Column</code></td>
<td>
<p>Manually specify effect allele column; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Upper_CI_Column">Upper_CI_Column</code></td>
<td>
<p>Manually specify upper CI column, when Pre_Calculated_CIs is TRUE; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Lower_CI_Column">Lower_CI_Column</code></td>
<td>
<p>Manually specify lower CI column, when Pre_Calculated_CIs is TRUE; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Left_Size">Left_Size</code></td>
<td>
<p>Manually specify the size of the grid text on the left hand side of the main plot; defaults to 9</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Right_Size">Right_Size</code></td>
<td>
<p>Manually specify the size of the grid text on the right hand side of the main plot; defaults to 9</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Right_Buffer">Right_Buffer</code></td>
<td>
<p>Manually specify the size of space around the grid text on the right hand side of the main plot; defaults to 1</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Left_Buffer">Left_Buffer</code></td>
<td>
<p>Manually specify the size of space around the grid text on the left hand side of the main plot; defaults to 2</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Axis_Buffer">Axis_Buffer</code></td>
<td>
<p>Manually specify the size of space around the axis end positions of the main plot; defaults to 0.02</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Left_Fill">Left_Fill</code></td>
<td>
<p>Manually specify the colour of the grid text background on the left hand side of the main plot; defaults to &quot;white&quot;</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Right_Fill">Right_Fill</code></td>
<td>
<p>Manually specify the colour of the grid text background on the right hand side of the main plot; defaults to &quot;white&quot;</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Grid_Colour_Left">Grid_Colour_Left</code></td>
<td>
<p>Manually specify the colour of the grid lines on the left hand side of the main plot; defaults to &quot;black&quot;</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Grid_Colour_Right">Grid_Colour_Right</code></td>
<td>
<p>Manually specify the colour of the grid lines on the right hand side of the main plot; defaults to &quot;black&quot;</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_CI_Label">CI_Label</code></td>
<td>
<p>Allow the display of a column containing CI information in the grid text module on the right hand side of the plot; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_BETA_Label">BETA_Label</code></td>
<td>
<p>Allow the display of a column containing BETA/OR information, depending on Test_Statistic_Choice in the grid text module on the right hand side of the plot; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_SE_Label">SE_Label</code></td>
<td>
<p>Allow the display of a column containing SE information in the grid text module on the right hand side of the plot; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_BETA_SE_Label">BETA_SE_Label</code></td>
<td>
<p>Allow the display of a column containing BETA &amp; SE information in the grid text module on the right hand side of the plot; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_BETA_CI_Label">BETA_CI_Label</code></td>
<td>
<p>Allow the display of a column containing BETA &amp; CI information in the grid text module on the right hand side of the plot; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_P_Label">P_Label</code></td>
<td>
<p>Allow the display of a column containing P-Value information in the grid text module on the right hand side of the plot; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Special_P">Special_P</code></td>
<td>
<p>Format the column containing P-Value information in the grid text module on the right hand side of the plot when P_Label is TRUE and P_Label is in order() to display as * or NS; defaults to FALSE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_CI_Label_Resolution">CI_Label_Resolution</code></td>
<td>
<p>Specify the number of decimal places of the column containing CI information in the grid text module on the right hand side of the plot; defaults to 3</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_BETA_Label_Resolution">BETA_Label_Resolution</code></td>
<td>
<p>Specify the number of decimal places of the column containing BETA information in the grid text module on the right hand side of the plot; defaults to 3</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_SE_Label_Resolution">SE_Label_Resolution</code></td>
<td>
<p>Specify the number of decimal places of the column containing SE information in the grid text module on the right hand side of the plot; defaults to 3</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_P_Label_Resolution">P_Label_Resolution</code></td>
<td>
<p>Specify the number of decimal places of the column containing P-Value information in the grid text module on the right hand side of the plot; defaults to 2</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_CI_Label_Title">CI_Label_Title</code></td>
<td>
<p>Specify the title of the column containing CI information in the grid text module on the right hand side of the plot; defaults to &quot;CI&quot;</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_BETA_Label_Title">BETA_Label_Title</code></td>
<td>
<p>Specify the title of the column containing BETA information in the grid text module on the right hand side of the plot; defaults to NULL</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_SE_Label_Title">SE_Label_Title</code></td>
<td>
<p>Specify the title of the column containing SE information in the grid text module on the right hand side of the plot; defaults to &quot;SE&quot;</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_BETA_SE_Label_Title">BETA_SE_Label_Title</code></td>
<td>
<p>Specify the title of the column containing SE information in the grid text module on the right hand side of the plot; defaults to NULL</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_BETA_CI_Label_Title">BETA_CI_Label_Title</code></td>
<td>
<p>Specify the title of the column containing BETA &amp; CI information in the grid text module on the right hand side of the plot; defaults to NULL</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_P_Label_Title">P_Label_Title</code></td>
<td>
<p>Specify the title of the column containing P information in the grid text module on the right hand side of the plot; defaults to &quot;P&quot;</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_order">order</code></td>
<td>
<p>Specify the order and selection of information containing columns (Labels) in the grid text module on the right hand side of the plot; defaults to c(&quot;BETA_SE_Label&quot;, &quot;P_Label&quot;)</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Verbose">Verbose</code></td>
<td>
<p>Prevent display of progress bar as function is running and instead show key milestone outputs/messages (mainly for debugging purposes); defaults to FALSE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Lab_Col">Lab_Col</code></td>
<td>
<p>Allocate the second level of information in the grid text module on the left hand side of the plot, when Double_Label is TRUE; defaults to &quot;Real_ID&quot;</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Unprocessed">Unprocessed</code></td>
<td>
<p>When using a raw model object, internal Model_Munge() processing to data frame is used; defaults to FALSE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_PDF">PDF</code></td>
<td>
<p>Slightly alter the underlying construction, suitable for PDF output; defaults to FALSE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_REF">REF</code></td>
<td>
<p>When supplying model data, toggle the REF: notation on or off; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_Top_Grid_Thickness">Top_Grid_Thickness</code></td>
<td>
<p>Modify the linewidth of the line at the very top of the grid, due to the potential graphics output; defaults to NULL, leading to automatic assignment unless otherwise specified</p>
</td></tr>
<tr><td><code id="Forest_Plot_:_HLA_Input">HLA_Input</code></td>
<td>
<p>Slightly alter the underlying processing, suitable for HLA-WAS inputs; defaults to FALSE</p>
</td></tr>
</table>


<h3>Value</h3>

<p>Image of Forest Plot is allocated to specified object and the resulting ggplot object can then be saved to an image
</p>


<h3>Examples</h3>

<pre><code class='language-R'>Forest_Plot &lt;- Forest_Plot(Data = "Intelligence_Sum_Stats", Chromosomes = 1)
</code></pre>


</div>
</div></div>
</body></html>
