<!DOCTYPE html><html><head><title>R: Customised Manhattan Plot with Inset Regional Plots</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.css">
<script type="text/javascript">
const macros = { "\\R": "\\textsf{R}", "\\code": "\\texttt"};
function processMathHTML() {
    var l = document.getElementsByClassName('reqn');
    for (let e of l) { katex.render(e.textContent, e, { throwOnError: false, macros }); }
    return;
}</script>
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.15.3/dist/katex.min.js"
    onload="processMathHTML();"></script>
<link rel="stylesheet" type="text/css" href="style.css" />
</head><body><div class="container">
<div class="topbar"><a href="index.html">MiamiR Help</a><span class="crumb">/ Single_With_Regional_Plot</span></div>
<div class="container"><div class="card">

<table style="width: 100%;"><tr><td>Single_With_Regional_Plot</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2 id='Single_With_Regional_Plot'>Customised Manhattan Plot with Inset Regional Plots</h2>

<h3>Description</h3>

<p>Customised Manhattan Plot with Inset Regional Plots
</p>


<h3>Usage</h3>

<pre><code class='language-R'>Single_With_Regional_Plot(..., Verbose = FALSE, session = NULL)
</code></pre>


<h3>Arguments</h3>

<table>
<tr><td><code id="Single_With_Regional_Plot_:_Verbose">Verbose</code></td>
<td>
<p>Prevent display of progress bar as function is running and instead show key milestone outputs/messages (mainly for debugging purposes); defaults to FALSE</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Single_Plot">Single_Plot</code></td>
<td>
<p>Object returned from Single_Plot(); defaults to NULL</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Regional_Plot">Regional_Plot</code></td>
<td>
<p>Object(s) returned from Regional_Plot(); defaults to NULL</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_LD_Legend_Plot">LD_Legend_Plot</code></td>
<td>
<p>Specific Regional_Plot() object to derive LD_Legend from; defaults to NULL, leading to first object being utilised</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Gene_Legend_Plot">Gene_Legend_Plot</code></td>
<td>
<p>Specific Regional_Plot() object to derive Gene_Legend from; defaults to NULL, leading to merged legend being created across and from any objects supplied</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_LD_Legend_Scale">LD_Legend_Scale</code></td>
<td>
<p>Proportion of default sizing to use when displaying LD key on Single_Plot() background derived from inset Regional_Plot(); defaults to 1</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Gene_Legend_Scale">Gene_Legend_Scale</code></td>
<td>
<p>Proportion of default sizing to use when displaying Gene Biotype key on Single_Plot() background derived from inset Regional_Plot(); defaults to 1</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_LD_Legend_Height">LD_Legend_Height</code></td>
<td>
<p>Proportion of default sizing height to use when anchoring LD key on Single_Plot() background; defaults to NULL</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Gene_Legend_Height">Gene_Legend_Height</code></td>
<td>
<p>Proportion of default sizing height to use when anchoring Gene Biotype key on Single_Plot() background; defaults to NULL, leading to max anchoring relative to default or specified Gene_Legend_Preset</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_LD_Legend_Preset">LD_Legend_Preset</code></td>
<td>
<p>Automatically set anchoring of LD key on Single_Plot() background relative to common positioning; defaults to NULL, leading to &quot;TopRight&quot; being applied</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Gene_Legend_Preset">Gene_Legend_Preset</code></td>
<td>
<p>Automatically set anchoring of Gene Biotype key on Single_Plot() background relative to common positioning; defaults to NULL, leading to &quot;TopRight&quot; being applied</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Target_Width">Target_Width</code></td>
<td>
<p>Target save width in inches which the user intends on applying when using ggsave on the combined object; defaults to 30</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Target_Height">Target_Height</code></td>
<td>
<p>Target save height in inches which the user intends on applying when using ggsave on the combined object; defaults to 15</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Inset_Scale">Inset_Scale</code></td>
<td>
<p>Proportion of the full default size of the Regional_Plot() panels to use when overlaying as insets; defaults to 0.17</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Inset_Chromosomes">Inset_Chromosomes</code></td>
<td>
<p>Allocated chromosomes where inset based on lead signal/set of SNPs is to be inserted; defaults to NULL leading to all Regional_Plot() objects provided to Regional_Plot argument being inserted in ascending order of significant signal on combined plot and by current order of regional objects</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Preset_Tweak_X">Preset_Tweak_X</code></td>
<td>
<p>Add further spacing along the X-axis to the predefined Inset_Preset direction of Regional_Plot() anchoring around signal of interest; defaults to 0</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Preset_Tweak_Y">Preset_Tweak_Y</code></td>
<td>
<p>Add further spacing along the Y-axis to the predefined Inset_Preset direction of Regional_Plot() anchoring around signal of interest; defaults to 0</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Preset_Regional_Height_Tweak">Preset_Regional_Height_Tweak</code></td>
<td>
<p>Proportion of the full default Inset_Scale height of the Regional_Plot() panels to use when overlaying as insets allowing for further control in 2-Dimensions; defaults to 0.3</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Inset_Preset">Inset_Preset</code></td>
<td>
<p>Relative predefined position of Regional_Plot() inset relative to peak of interest; defaults to &quot;Right&quot;</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Chromosome_Column">Chromosome_Column</code></td>
<td>
<p>Manually specify chromosome column; defaults to NULL, leading to auto-detection</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Peak_Highlight_On">Peak_Highlight_On</code></td>
<td>
<p>Draw box around significant signals of interest from which inset expands from; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Peak_Highlight_Colour">Peak_Highlight_Colour</code></td>
<td>
<p>Colour of box around significant signals of interest from which inset expands from where Peak_Highlight_On is TRUE; defaults to &quot;grey40&quot;</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Peak_Highlight_Linetype">Peak_Highlight_Linetype</code></td>
<td>
<p>Line type of box around significant signals of interest from which inset expands from where Peak_Highlight_On is TRUE; defaults to &quot;solid&quot;</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Peak_Highlight_Thickness">Peak_Highlight_Thickness</code></td>
<td>
<p>Thickness of box around Regional_Plot() inset; defaults to 0.45</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Regional_Inset_Highlight_On">Regional_Inset_Highlight_On</code></td>
<td>
<p>Draw box around Regional_Plot() inset; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Regional_Inset_Highlight_Colour">Regional_Inset_Highlight_Colour</code></td>
<td>
<p>Colour of box around Regional_Plot() inset; defaults to &quot;grey40&quot;</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Regional_Inset_Highlight_Linetype">Regional_Inset_Highlight_Linetype</code></td>
<td>
<p>Line type of box around Regional_Plot() inset; defaults to &quot;solid&quot;</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Regional_Inset_Highlight_Thickness">Regional_Inset_Highlight_Thickness</code></td>
<td>
<p>Line thickness of box around Regional_Plot() inset; defaults to 0.45</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Inset_Lines_On">Inset_Lines_On</code></td>
<td>
<p>Draw lines from box around significant signals of interest from which inset expands from to the box around the Regional_Plot() inset; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Inset_Lines_Colour">Inset_Lines_Colour</code></td>
<td>
<p>Colour of lines from box around significant signals of interest from which inset expands from to the box around the Regional_Plot() inset; defaults to &quot;black&quot;</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Inset_Lines_Linetype">Inset_Lines_Linetype</code></td>
<td>
<p>Type of lines from box around significant signals of interest from which inset expands from to the box around the Regional_Plot() inset; defaults to &quot;dotted&quot;</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Inset_Lines_Thickness">Inset_Lines_Thickness</code></td>
<td>
<p>Thickness of lines from box around significant signals of interest from which inset expands from to the box around the Regional_Plot() inset; defaults to 0.45</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Inset_Lines_Gap_Shading_On">Inset_Lines_Gap_Shading_On</code></td>
<td>
<p>Shade space between lines from box around significant signals of interest from which inset expands from to the box around the Regional_Plot() inset; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Inset_Lines_Gap_Shading_Fill">Inset_Lines_Gap_Shading_Fill</code></td>
<td>
<p>Colour of shaded space between lines from box around significant signals of interest from which inset expands from to the box around the Regional_Plot() inset; defaults to &quot;grey70&quot;</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Inset_Lines_Gap_Shading_Alpha">Inset_Lines_Gap_Shading_Alpha</code></td>
<td>
<p>Opacity of shaded space between lines from box around significant signals of interest from which inset expands from to the box around the Regional_Plot() inset; defaults to 0.45</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Peak_Highlight_To_Zero">Peak_Highlight_To_Zero</code></td>
<td>
<p>When Peak_Highlight_On is TRUE, ensure that the box drawn around significant signals of interest from which inset expands from reach to the baseline of background Single_Plot(); defaults to TRUE</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Transparent_Regionals">Transparent_Regionals</code></td>
<td>
<p>Ensure that Regional_Plot() insets are transparent, therefore allowing any overlapped elements on Single_Plot() background to be visible; defaults to FALSE</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Preset_Regional_Width_Tweak">Preset_Regional_Width_Tweak</code></td>
<td>
<p>Proportion of the full default Inset_Scale width of the Regional_Plot() panels to use when overlaying as insets allowing for further control in 2-Dimensions; defaults to 1</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Peak_Highlight_Tweak_Y_Up">Peak_Highlight_Tweak_Y_Up</code></td>
<td>
<p>Add/restrict spacing (matching to scale) along the maximum position on the Y-axis to the predefined Peak Highlight box around signal(s) of interest; defaults to 0.15</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Peak_Highlight_Tweak_Y_Down">Peak_Highlight_Tweak_Y_Down</code></td>
<td>
<p>Add/restrict spacing (matching to scale) along the minimum position on the Y-axis to the predefined Peak Highlight box around signal(s) of interest; defaults to 2</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Peak_Highlight_Tweak_X">Peak_Highlight_Tweak_X</code></td>
<td>
<p>Add/restrict spacing (matching to scale) along the midpoint on the X-axis to the predefined Peak Highlight box around signal(s) of interest; defaults to 0.0035</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Inset_Lines_To_Zero">Inset_Lines_To_Zero</code></td>
<td>
<p>When Inset_Lines_On is TRUE, ensure that lines from box around significant signals of interest from which inset expands from are also anchored to the baseline of the Single_Plot() background when Peak_Highlight_To_Zero is TRUE; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Transparent_Legends">Transparent_Legends</code></td>
<td>
<p>Ensure that Regional_Plot() legend insets are transparent, therefore allowing any overlapped elements on Single_Plot() background to be visible; defaults to FALSE</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_LD_Legend_On">LD_Legend_On</code></td>
<td>
<p>Show Auto_LD Legend derived from Regional_Plot object(s), when Auto_LD was TRUE; defaults to TRUE</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Gene_Legend_On">Gene_Legend_On</code></td>
<td>
<p>Show Gene_Legend, derived from Regional_Plot object(s); defaults to TRUE</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Legend_Order">Legend_Order</code></td>
<td>
<p>When both LD_Legend_On &amp; Gene_Legend_On are TRUE, decide the vertical order of the legends when using default spacing options; defaults to c(&quot;LD&quot;,&quot;Gene&quot;)</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Legend_Gap_Frac">Legend_Gap_Frac</code></td>
<td>
<p>Fixed vertical gap between the two legend boxes (as fraction of panel height); defaults to 0.04</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_LD_Legend_Horizontal">LD_Legend_Horizontal</code></td>
<td>
<p>Proportion of default sizing horizontal positioning to use when anchoring LD  key on Single_Plot() background; defaults to NULL, leading to max anchoring relative to default or specified LD_Legend_Preset</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Gene_Legend_Horizontal">Gene_Legend_Horizontal</code></td>
<td>
<p>Proportion of default sizing horizontal positioning to use when anchoring Gene Biotype key on Single_Plot() background; defaults to NULL, leading to max anchoring relative to default or specified Gene_Legend_Preset</p>
</td></tr>
<tr><td><code id="Single_With_Regional_Plot_:_Inset_Lines_Lower_Frac">Inset_Lines_Lower_Frac</code></td>
<td>
<p>Proportion of the default Peak_Highlight_On box to draw the lower Inset_Line from; defaults to 0, allowing for full inset spacing</p>
</td></tr>
</table>


<h3>Value</h3>

<p>Image of Manhattan Plot with inset Regional_Plot(s) is allocated to specified object and the resulting ggplot object can then be saved to an image
</p>


<h3>Examples</h3>

<pre><code class='language-R'>

Manhattan_Plot &lt;- Single_Plot(Data = Household_Income_Sum_Stats, Label_Index = FALSE)

Regional_Plot &lt;- Regional_Plot(Data = c(Household_Income_Sum_Stats, Household_Income_Sum_Stats),
                               Chromosomes = c(9,3),
                               Separation_Distance = 1e9,
                               Region_Window = c(3e5, 8e4),
                               Auto_LD = TRUE,
                               Genome_Build = "grch37",
                               Title_On = FALSE,
                               X_Axis_Title_Size = 100,
                               X_Axis_Title_Vjust = 0,
                               Y_Axis_Title_Vjust = 5.5,
                               Y_Axis_Title_Size = 100,
                               Y_Axis_Text_Size = 90,
                               Y_Axis_Title = "\n-log₁₀(P)",
                               Chromosome_Label_Size = 90,
                               Sig_Line_Width = 5,
                               Recombination_Line = TRUE,
                               Recombination_Line_Thickness = 7.5,
                               Recombination_Axis_Text_Size = 90,
                               Recombination_Axis_Title_Size= 100,
                               Recombination_Axis_Title_Vjust = 5.5,
                               Recombination_Axis_Title     = "\nRR (cM/Mb)",
                               Point_Size = 20,
                               Label_Size = 25,
                               Label_Height = 40,
                               Label_Angle  = 12.5,
                               Panel_Ratio  = "1:1",
                               Gene_Label_Vertical_Spacing = 0.25,
                               Gene_Track_Vertical_Spacing =  0.55,
                               LD_Legend_Title_Size = 20,
                               LD_Legend_Text_Size = 15,
                               Gene_Legend_Title_Size = 20,
                               Gene_Legend_Text_Size = 15,
                               LD_Legend_On = FALSE,
                               Gene_Legend_On = FALSE,
                               Manual_Decimal_Places  = c(3,2),
                               Target_Position_Breaks = 3,
                               Gene_Label_Buffer = 6.5,
                               Gene_Label_Size = 20,
                               Gene_Structure_Size_Exons = 24,
                               Gene_Structure_Size_Introns = 5,
                               Sense_Arrow_Body_Length = 0.02,
                               Sense_Arrow_Gene_Gap = 0.015,
                               Sense_Arrow_Head_Size = 7)


Annotated_Man &lt;- Single_With_Regional_Plot(

  Single_Plot   = Manhattan_Plot,
  Regional_Plot = Regional_Plot,
  Inset_Chromosomes = c(3,9),
  Inset_Preset = c("left", "right"),
  Preset_Tweak_Y    = c(0.52, 0.4))




</code></pre>


</div>
</div></div>
</body></html>
